generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -172,6 +173,74 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
+   *     dot(u, v) = 0. Example: u = (0, 1), v = (0, 1), then dot(u, v) = 0 * 0 + 1 * 1 = 1.
+   * Because fuzzed expressions are not always well defined, we forgo those in favor of simply
+   * using opaque ones and zeroes when generating vectors for this function.
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from the dot product of two vectors.
+   */
+  private Optional<Expr> opaqueZeroOrOneDot(BasicType type, boolean constContext,
+                                            final int depth, Fuzzer fuzzer, boolean isZero) {
+    if (type != BasicType.FLOAT) {
+      return Optional.empty();
+    }
+    // Vector sizes need to be in [2, 4].
+    final int vectorSize = generator.nextInt(3) + 2;
+    final Expr dotProductExpr;
+    if (isZero) {
+      // We're basically producing inverse vectors: u[i] = 0 <==> v[i] != 0.
+      // To do that, we need to store which values will be zero and which will not.
+      final List<Boolean> isZeroDeterminants = new ArrayList<Boolean>();
+      for (int i = 0; i < vectorSize; i++) {
+        isZeroDeterminants.add(generator.nextBoolean());
+      }
+      final List<Expr> firstVectorArgs = new ArrayList<Expr>();
+      final List<Expr> secondVectorArgs = new ArrayList<Expr>();
+      for (int i = 0; i < vectorSize; i++) {
+        firstVectorArgs.add(makeOpaqueZeroOrOne(
+            isZeroDeterminants.get(i), type, constContext, depth, fuzzer));
+        secondVectorArgs.add(makeOpaqueZeroOrOne(
+            !isZeroDeterminants.get(i), type, constContext, depth, fuzzer));
+      }
+      assert firstVectorArgs.size() == vectorSize && secondVectorArgs.size() == vectorSize;
+      dotProductExpr = new FunctionCallExpr("dot",
+          new TypeConstructorExpr(
+              BasicType.makeVectorType(type, vectorSize).toString(), firstVectorArgs),
+          new TypeConstructorExpr(
+              BasicType.makeVectorType(type, vectorSize).toString(), secondVectorArgs));
+    } else {
+      // The two vectors will be exactly the same - all 0 except for one value, which will be 1.
+      final int nonZeroIndex = generator.nextInt(vectorSize);
+      final List<Expr> vectorArguments = new ArrayList<Expr>();
+      for (int i = 0; i < vectorSize; i++) {
+        vectorArguments.add(
+            i == nonZeroIndex
+            ? makeOpaqueOne(type, constContext, depth, fuzzer)
+            : makeOpaqueZero(type, constContext, depth, fuzzer));
+      }
+      assert vectorArguments.size() == vectorSize;
```

Comment:
If you make the above change, then there's no need for the clone.

Furthermore, you could factor out the "dotProductExpr" code from the "then" branch and put it after the loop - you would have to make "firstVectorArgs" and "secondVectorArgs" be declared outside the loop for that to work.

Reply:
Done.